### ISSUE TITLE
Errors while provisioning with NFS leader 

### ISSUE THREAD

**Issue by careermonk** [kubernetes-incubator/external-storage#43]
We are using NFS provisioner for dynamic provisioning. While provisioning a claim, are we allowed to make changes to cluster? Can we add a worker/delete a node from cluster? If we do that, the leader is failing with below error.

We are seeing the following errors in storage pod:

```E0328 15:59:11.926801       1 leaderelection.go:245] error initially creating leader election record: create not allowed, PVC should already exist!
E0328 15:59:13.926741       1 leaderelection.go:245] error initially creating leader election record: create not allowed, PVC should already exist!
E0328 15:59:15.927195       1 leaderelection.go:245] error initially creating leader election record: create not allowed, PVC should already exist!```

**Status logs:**

```
Mar 28 15:58:47.913: INFO: Waiting up to 15m0s for PersistentVolumeClaim pvc-gq10r to have phase Bound
Mar 28 15:58:47.915: INFO: PersistentVolumeClaim pvc
Mar 28 15:58:57.917: INFO: PersistentVolumeClaim pvc-gq10r found but phase is Pending instead of Bound.
Mar 28 15:59:07.920: INFO: PersistentVolumeClaim pvc-gq10r found but phase is Pending instead of Bound.
Mar 28 15:59:17.921: INFO: Get persistent volume claim pvc-gq10r in failed, ignoring for 10s: persistentvolumeclaims "pvc-gq10r" not found
Mar 28 15:59:27.923: INFO: Get persistent volume claim pvc-gq10r in failed, ignoring for 10s: persistentvolumeclaims "pvc-gq10r" not found
Mar 28 15:59:37.925: INFO: Get persistent volume claim pvc-gq10r in failed, ignoring for 10s: persistentvolumeclaims "pvc-gq10r" not found
```

**Comment by wongma7** [comment]
Do the leaderelection errors eventually stop? They should stop after ~15 seconds and you can safely ignore them.

You are allowed to delete a claim before the NFS provisioner has created a PV for it, nothing too bad should happen. The worst-case scenario is: the PV gets created right after you delete the PVC and then you are left with a dangling PV which cannot be claimed by any PVC except the one you just deleted. But you can easily clean up those PVs. Does this answer your questions?

**Comment by careermonk** [comment]
Yes it does. We have kept retry logic in our code. Once leader starts giving FAILED state it not giving correct state even though the PVC state is BOUND, it is giving the reply as FAILED.

**Comment by careermonk** [comment]
So leader is not getting correct state once it gives FAILED state. In the logs, I see it got successfully provisioned.

But the status is showing wrongly.

```
{"level":"info","ts":"2017-03-28T16:03:38Z","msg":"file_storage.go:123: Storage created but waiting for Active Transactions completion","OrderId":13452467,"OrderId":21918905,"Active Transactions":1}
{"level":"info","ts":"2017-03-28T16:03:55Z","msg":"file_storage.go:135: Successfully created Storage","StorageId":21918905,"Size":40,"IOPS":2,"Location":"DC01"}
```

**Comment by careermonk** [comment]
Can we add a worker/delete a worker from cluster? Does the code can handle that?

**Comment by wongma7** [comment]
Sorry I don't totally understand, what do you mean by worker? & what is giving FAILED state? 

Let me sum up what the leaderelection.go code does since it is undocumented and its name is misleading... The leaderelection code is for when you run multiple controller-provisioners and you want only one of them to call `Provision` on a PVC. All the provisioners will race-to-lock/"become the leader" of the PVC and only the leader will call `Provision.`

This error "error initially creating leader election record: create not allowed, PVC should already exist!" means somehow, a provisioner became the leader of the PVC but then the PVC got deleted from the cluster. So it will keep posting that error until it realizes the PVC has been deleted.

**Comment by careermonk** [comment]
@wongma7 We are running the provisioner in a a cluster. Let us say PVC provision is in progress with one of the worker node in that cluster. If we delete that worker while the PVC provision is in progress, we are seeing the failures. Is that a valid scenario?

**Comment by wongma7** [comment]
Yes if I understand you correctly it is valid. If you delete the worker-PVC "pvc-gq10r" while provisioning is in progress: 
* The leader provisioner will fail to renew the lease on the PVC (report "error initially creating leader election record: create not allowed, PVC should already exist!" once) BUT it may still create the (dangling) PV because provsioning is already in progress and cannot be stopped
* Other non-leader provisioners, if you are running any, will fail to acquire the lease on the PVC (report "error initially creating leader election record: create not allowed, PVC should already exist!" repeatedly) and eventually will give up after ~15s

 In short, this error is expected if you delete the PVC while provisioning is in progress. Full logs & error messages may help me understand the scenario better 👍

**Comment by midoblgsm** [comment]
@wongma7 what happens if the node that contains the leader provisioner is deleted before the PVC is bound?

**Comment by wongma7** [comment]
* If the provisioner is killed after it has created the storage asset but
before it has created the PV then you are left with a dangling storage
asset. I assume that's what's meant by

{"level":"info","ts":"2017-03-28T16:03:55Z","msg":"file_storage.go:135:
Successfully created
Storage","StorageId":21918905,"Size":40,"IOPS":2,"Location":"DC01"}

I don't really see a way of the provisioner recognizing a dangling
storage asset and cleaning it up for you, but maybe it's possible?

* If the provisioner was killed before it created the storage asset,
no problem, when it restarts it will try again

* If it was killed right after it created the storage asset and after
the PV, no problem either because the PVC will get bound to that PV


Sorry for my confusion, the logs posted refer to a PVC being deleted
but the case of node getting deleted/drained is something else

**Comment by wongma7** [comment]
Hm, I think I am also wrongly equating "killing the pod" to "deleting the node." 

If the provisioner loses contact with kubernetes (because the node has been deleted) BUT it is still running in the middle of Provision, it will try and fail repeatedly to create the Kubernetes PV. After a certain number of failures because it cannot contact kubernetes, it *will* Delete the storage asset.

Does this answer your concerns?

**Comment by midoblgsm** [comment]
My concern is: when the provisioner picks up the PVC,  the PVC state will be pending and the provisioner gets the lock on that PVC (correct me if I am wrong). Then, when the provisioner node dies (let's say that the node will not recover and it will be lost), do we have any mechanism to force release the lock on that PVC so any other running provisioner can pick it up (after a given time) ? or we will have that PVC in pending state until someone deletes it?

**Comment by wongma7** [comment]
Yes, it's the last parameter to the controller, "termlimit" I believe which is the max timeout a provisioner has to succeed/fail before it is forced to give up the lock . So If a provisioner gets lost, other provisioners in the cluster will be able to see how long it has been since the lost provisioner updated the lock, and if the lost provisioner doesn't update in time/when the timeout is reached they will race to get the pvc again

**Comment by careermonk** [comment]
I see the provisioning was successful but the status of it was showing failed with this scenario.

**Comment by wongma7** [comment]
where did these messages "Mar 28 15:59:37.925: INFO: Get persistent volume claim pvc-gq10r in failed, ignoring for 10s: persistentvolumeclaims "pvc-gq10r" not found" come from? What status shows failed: do you mean events on the PV, events on the PVC, or somewhere else? By events I mean the list of things that show when you do "kubectl describe"

**Comment by careermonk** [comment]
Data us coming from e2e tests.

**Comment by wongma7** [comment]
Okay. Can you please walk me through the exact steps you did to find this failure so I can reproduce it? 

I enumerated all the cases where a node gets deleted, none can result in a pvc getting deleted like your logs show

**Comment by wongma7** [comment]
closing; I've changed the e2e tests a bit since this
